{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import random\n",
    "import copy\n",
    "\n",
    "import re\n",
    "from matplotlib import cm\n",
    "import matplotlib.patches as patches\n",
    "\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "import scipy.stats\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sequence_df = pd.read_csv('preprocess_datasets/preprocessed_datasets/2019_guo_nci60_formatted_peptide_quants.tsv',\n",
    "                          sep = '\\t', index_col = 0)\n",
    "sequence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_proteins = sequence_df['Protein'].values\n",
    "all_unique_proteins = np.unique(sequence_df['Protein'].values)\n",
    "print(\"All proteins: \", len(all_unique_proteins))\n",
    "print(\"All proteins: \", all_unique_proteins)\n",
    "\n",
    "all_peptide_sequences = sequence_df['Peptide']\n",
    "\n",
    "print(\"All sequences: \", len(all_peptide_sequences))\n",
    "print(\"All sequences: \", all_peptide_sequences[:10])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download all the protein sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Entrez.email = \"abdincer@uw.edu\"  # Always tell NCBI who you are\n",
    "\n",
    "#Record protein sequences\n",
    "all_protein_sequences = []\n",
    "all_protein_information = []\n",
    "\n",
    "for protein_id in all_unique_proteins:\n",
    "    \n",
    "    filename = \"protein_fasta_files/\" + protein_id + \".fasta\"\n",
    "    if not os.path.isfile(filename):\n",
    "        # Downloading...\n",
    "        net_handle = Entrez.efetch(\n",
    "            db=\"protein\", id=protein_id, rettype=\"fasta\", retmode=\"text\"\n",
    "        )\n",
    "        out_handle = open(filename, \"w\")\n",
    "        out_handle.write(net_handle.read())\n",
    "        out_handle.close()\n",
    "        net_handle.close()\n",
    "        print(\"Saved\")\n",
    "\n",
    "    print(\"Parsing...\")\n",
    "    record = SeqIO.read(filename, \"fasta\")\n",
    "    print(record)\n",
    "    all_protein_sequences.append(record.seq)\n",
    "    all_protein_information.append(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Match peptide sequences with proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Record the indices for each peptide\n",
    "peptide_sequence_start_indices = np.zeros(len(all_peptide_sequences))\n",
    "peptide_sequence_end_indices = np.zeros(len(all_peptide_sequences))\n",
    "\n",
    "for p_index, protein in enumerate(all_unique_proteins):\n",
    "    \n",
    "    protein_sequence = all_protein_sequences[p_index]\n",
    "    print(\"Protein sequence: \", protein)\n",
    "    print(\"Protein sequence: \", protein_sequence)\n",
    "    \n",
    "    peptides = all_peptide_sequences.iloc[np.where(all_proteins == protein)[0]]\n",
    "    print(peptides)\n",
    "    for peptide_index, peptide in enumerate(peptides):\n",
    "        \n",
    "        print(\"Peptide sequence:\", peptide)\n",
    "        print(\"Peptide sequence:\", peptide_index)\n",
    "        try:\n",
    "            index_order = protein_sequence.index(peptide)\n",
    "            print(\"Index of peptide:\", index_order)\n",
    "        except:\n",
    "            print(\"Not detected\")\n",
    "        \n",
    "        #Record the results\n",
    "        print([np.where(sequence_df.index == peptides.index[peptide_index])[0]])\n",
    "        peptide_sequence_start_indices[np.where(sequence_df.index == peptides.index[peptide_index])[0]] = index_order\n",
    "        peptide_sequence_end_indices[np.where(sequence_df.index == peptides.index[peptide_index])[0]] = index_order + len(peptide)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_df['Peptide start index'] = peptide_sequence_start_indices\n",
    "sequence_df['Peptide end index'] = peptide_sequence_end_indices\n",
    "\n",
    "#Record results\n",
    "peptide_indices = sequence_df[['Peptide', 'Protein', 'gene symbol', 'R2', 'best mscore',\n",
    "       'numNA', 'Charge 2', 'Charge 3', 'Charge 4', 'Charge 5', 'Charge 6', \n",
    "                          'Peptide start index', 'Peptide end index']]\n",
    "\n",
    "peptide_indices.index = sequence_df.index\n",
    "peptide_indices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read abundance matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_filename = 'trained_models/2019_guo_nci60/2019_guo_nci60_'\n",
    "\n",
    "#Read Q, alpha, and c matrices\n",
    "Q_train = pd.read_csv(output_filename + 'Q_train.tsv', sep = '\\t', index_col = 0)\n",
    "Q_test = pd.read_csv(output_filename + 'Q_test.tsv', sep = '\\t', index_col = 0)\n",
    "\n",
    "P_train = pd.read_csv(output_filename + 'P_train.tsv', sep = '\\t', index_col = 0)\n",
    "P_test = pd.read_csv(output_filename + 'P_test.tsv', sep = '\\t', index_col = 0)\n",
    "\n",
    "predicted_train_coeffs = pd.read_csv(output_filename + 'coefficients_train.tsv', sep = '\\t', index_col = 0)\n",
    "predicted_test_coeffs = pd.read_csv(output_filename + 'coefficients_test.tsv', sep = '\\t', index_col = 0)\n",
    "predicted_train_coeffs.index = Q_train.index\n",
    "predicted_test_coeffs.index = Q_test.index\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createJointPlots(protein_name):\n",
    "    \n",
    "    ############################################\n",
    "    # Define plot shapes\n",
    "    SMALL_SIZE = 60\n",
    "    MEDIUM_SIZE = 60\n",
    "    BIGGER_SIZE = 90\n",
    "\n",
    "    plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "    plt.rc('axes', titlesize=MEDIUM_SIZE)     # fontsize of the axes title\n",
    "    plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "    plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "    plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "    fig, (ax1, ax2, ax3) = plt.subplots(3, 1, sharex = True, figsize=(40, 30))\n",
    "    fig.tight_layout()\n",
    "    \n",
    "    \n",
    "    ############################################\n",
    "    #We want to sort the peptides and then get the abundances\n",
    "    \n",
    "    sorted_peptide_indices = peptide_indices.loc[sequence_df[sequence_df['Protein'] == protein_name].index].sort_values(by = ['Peptide start index', 'Peptide end index']).index\n",
    "   \n",
    "    #Get sorted raw abundances\n",
    "    index_values = peptide_indices.loc[sorted_peptide_indices]['Peptide start index'].values\n",
    "    index_values_end = peptide_indices.loc[sorted_peptide_indices]['Peptide end index'].values\n",
    "    raw_peptide_abundances = Q_test.loc[sequence_df[sequence_df['Protein'] == protein_name].index].loc[sorted_peptide_indices]\n",
    "\n",
    "    mean_raw_abundance = np.mean(raw_peptide_abundances.values)\n",
    "    print(\"Mean raw abundance: \", mean_raw_abundance)\n",
    "    \n",
    "    #Plot the mean abundance across all runs for each peptide\n",
    "    raw_peptide_abundances_run_mean = np.mean(raw_peptide_abundances.values, axis = 1)\n",
    "   \n",
    "\n",
    "    #Obtain adjusted abundances using coefficients\n",
    "    sorted_peptide_indices = peptide_indices.loc[sequence_df[sequence_df['Protein'] == protein_name].index].sort_values(by = ['Peptide start index', 'Peptide end index']).index\n",
    "    raw_peptide_abundances = Q_test.loc[sequence_df[sequence_df['Protein'] == protein_name].index].loc[sorted_peptide_indices]\n",
    "\n",
    "    coeffs = predicted_test_coeffs.loc[sequence_df[sequence_df['Protein'] == protein_name].index].loc[sorted_peptide_indices].values.ravel().reshape((-1, 1))\n",
    "    adjusted_peptide_abundances = raw_peptide_abundances / np.abs(coeffs)\n",
    "    \n",
    "    mean_adjusted_abundance = np.mean(adjusted_peptide_abundances.values)\n",
    "    print(\"Mean adjusted abundance: \", mean_adjusted_abundance)\n",
    "    \n",
    "    #Plot the mean abundance across all runs for each peptide\n",
    "    adjusted_peptide_abundances_run_mean = np.mean(adjusted_peptide_abundances.values, axis = 1)\n",
    "    \n",
    "    \n",
    "    ############################################\n",
    "    #Center all the raw and observed abundances\n",
    "    \n",
    "    raw_peptide_abundances_run_mean = raw_peptide_abundances_run_mean - mean_raw_abundance\n",
    "    adjusted_peptide_abundances_run_mean = adjusted_peptide_abundances_run_mean - mean_adjusted_abundance\n",
    "    \n",
    "    mean_raw_abundance = np.mean(raw_peptide_abundances_run_mean)\n",
    "    mean_adjusted_abundance = np.mean(adjusted_peptide_abundances_run_mean)\n",
    "    \n",
    "    \n",
    "    ############################################\n",
    "    #Plot the raw peptide abundances\n",
    "    \n",
    "    #Plot abundances of all peptides\n",
    "    for i in range(len(raw_peptide_abundances_run_mean)):\n",
    "        peptide_charges = peptide_indices.loc[sorted_peptide_indices].iloc[i]\n",
    "          \n",
    "        ax1.vlines(x = (index_values[i] + index_values_end[i]) / 2, \n",
    "                       ymin = mean_raw_abundance, ymax = raw_peptide_abundances_run_mean[i],\n",
    "                   linewidth = 5, color = '#eb4d4b')\n",
    "        \n",
    "        ax1.scatter((index_values[i] + index_values_end[i]) / 2, raw_peptide_abundances_run_mean[i], alpha = 0.5, \n",
    "                    s = 1500, color = '#eb4d4b', zorder = 10)\n",
    "        \n",
    "    protein_sequence = all_protein_sequences[np.where(all_unique_proteins == protein_name)[0][0]]\n",
    "    \n",
    "    #Plot the mean raw peptide abundance\n",
    "    ax1.plot(np.arange(len(protein_sequence)), \n",
    "              mean_raw_abundance * np.ones(len(protein_sequence)), \n",
    "              lw = 5, color = 'Black', label = 'Mean observed abundance')\n",
    "\n",
    "    \n",
    "    ############################################\n",
    "    #Plot the adjusted peptide abundances\n",
    "    \n",
    "    for i in range(len(adjusted_peptide_abundances_run_mean)):\n",
    "        peptide_charges = peptide_indices.loc[sorted_peptide_indices].iloc[i]\n",
    "        \n",
    "        ax2.vlines((index_values[i] + index_values_end[i]) / 2, \n",
    "                   ymin = mean_adjusted_abundance, ymax = adjusted_peptide_abundances_run_mean[i],\n",
    "               linewidth = 5, color = '#8854d0')\n",
    "        ax2.scatter((index_values[i] + index_values_end[i]) / 2, \n",
    "                    adjusted_peptide_abundances_run_mean[i], alpha = 0.5, \n",
    "                s = 1500, color = '#8854d0', zorder = 10)\n",
    "        \n",
    "    ax2.plot(np.arange(len(protein_sequence)), \n",
    "              mean_adjusted_abundance * np.ones(len(protein_sequence)), \n",
    "              lw = 5, color = 'Black', label = 'Mean adjusted abundance')\n",
    "\n",
    "    \n",
    "    ############################################\n",
    "    #Plot the alignment of peptides along the protein, colored by charge \n",
    "    \n",
    "    current_height = 0\n",
    "    for charge in range(2, 7):\n",
    "        \n",
    "        #Plot each peptide as a bar plot\n",
    "        current_end_indices = [0]\n",
    "        available_heights = [0]\n",
    "\n",
    "        for i in range(len(raw_peptide_abundances_run_mean)):\n",
    "            peptide_charges = peptide_indices.loc[sorted_peptide_indices].iloc[i]\n",
    "\n",
    "            if peptide_charges['Charge 2'] == 1:\n",
    "                color_value = '#22a6b3'\n",
    "            elif peptide_charges['Charge 3'] == 1:\n",
    "                color_value = '#f0932b'\n",
    "            elif peptide_charges['Charge 4'] == 1:\n",
    "                color_value = '#f9ca24'\n",
    "            elif peptide_charges['Charge 5'] == 1:\n",
    "                color_value = '#be2edd'\n",
    "            elif peptide_charges['Charge 6'] == 1:\n",
    "                color_value = '#badc58'\n",
    "\n",
    "            if peptide_charges['Charge ' + str(charge)] == 1:\n",
    "                start_of_peptide = index_values[i]\n",
    "                end_of_peptide = index_values_end[i]\n",
    "\n",
    "                #Compare the start site to all heights\n",
    "                plotted_flag = False\n",
    "                for height in available_heights:\n",
    "                    if not plotted_flag and start_of_peptide > current_end_indices[height]:\n",
    "        \n",
    "                        rect = patches.Rectangle((index_values[i], current_height + (height/2)), \n",
    "                                      index_values_end[i] - index_values[i], 0.4,\n",
    "                                      facecolor = color_value, linewidth=2, \n",
    "                                      alpha = 0.5, edgecolor='black')\n",
    "                        current_end_index = end_of_peptide\n",
    "                        current_end_indices[height] = current_end_index\n",
    "                        plotted_flag = True\n",
    "                        ax3.add_patch(rect)\n",
    "                    \n",
    "                height = available_heights[-1]\n",
    "                if not plotted_flag:\n",
    "                    available_heights.append((height + 1))\n",
    "                    height = height + 1\n",
    "                    rect = patches.Rectangle((index_values[i], current_height + (height/2)), \n",
    "                                      index_values_end[i] - index_values[i], 0.4,\n",
    "                                      facecolor = color_value, linewidth=2, \n",
    "                                      alpha = 0.5, edgecolor='black')\n",
    "                    ax3.add_patch(rect)\n",
    "                    current_end_indices.append(end_of_peptide)\n",
    "\n",
    "            \n",
    "        #Increase height \n",
    "        current_height = current_height + (available_heights[-1]/2) + 0.5\n",
    "        \n",
    "        \n",
    "    ############################################\n",
    "    #Plot cleavage points\n",
    "\n",
    "    protein_sequence = all_protein_sequences[np.where(all_unique_proteins == protein_name)[0][0]]\n",
    "    for i in range(len(protein_sequence)-1):\n",
    "        if protein_sequence[i:i+1] == 'K' or protein_sequence[i:i+1] == 'R':\n",
    "            \n",
    "            ax3.vlines(x = i+1, \n",
    "                      ymin = -0.1, \n",
    "                      ymax = 0.0,\n",
    "                      linewidth = 2,\n",
    "                      color = 'black')\n",
    "    \n",
    "    \n",
    "    ############################################\n",
    "    #Define grids, labels, axes\n",
    "    \n",
    "    ax1.grid('minor')\n",
    "    ax2.grid('minor')\n",
    "    \n",
    "    max_value = max(np.nanmax(raw_peptide_abundances_run_mean), \n",
    "                    np.nanmax(adjusted_peptide_abundances_run_mean)) * 1.25\n",
    "    \n",
    "    min_value = min(np.nanmin(raw_peptide_abundances_run_mean), \n",
    "                    np.nanmin(adjusted_peptide_abundances_run_mean)) * 1.5\n",
    "    \n",
    "    ax1.set_ylim([min_value, max_value])\n",
    "    ax2.set_ylim([min_value, max_value])\n",
    "    \n",
    "    \n",
    "    ax3.set_xlabel('Peptide index of protein ' + protein_name)\n",
    "    ax1.set_ylabel('Centered \\n observed \\n abundances')\n",
    "    ax2.set_ylabel('Centered \\n adjusted \\n abundances')\n",
    "    ax3.set_ylabel('Peptides \\n along \\n the protein')\n",
    "    \n",
    "    raw_error = np.mean(np.abs(raw_peptide_abundances_run_mean - mean_raw_abundance)**2)\n",
    "    adj_error = np.mean(np.abs(adjusted_peptide_abundances_run_mean - mean_adjusted_abundance)**2)\n",
    "    percent_improvement = ((raw_error - adj_error) / raw_error) * 100\n",
    "    \n",
    "    #Count no of peptides that becomes closer to the mean\n",
    "    count_peptide = np.where(np.abs(adjusted_peptide_abundances_run_mean - mean_adjusted_abundance) < np.abs(raw_peptide_abundances_run_mean - mean_raw_abundance))[0]\n",
    "    print(\"No of peptides with smaller error: \", len(count_peptide))\n",
    "    print(\"Percent of peptides with smaller error: \", (len(count_peptide) / len(adjusted_peptide_abundances_run_mean)))\n",
    "    \n",
    "    print(\"Error for raw values: \", raw_error)\n",
    "    print(\"Error for adjusted values: \", adj_error)\n",
    "    print(\"Percent improvement: \", percent_improvement)\n",
    "    \n",
    "    plt.xlim([-5, len(protein_sequence) + 5]) \n",
    "\n",
    "    ax3.set_ylim([0, current_height])\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select one of the proteins containing many peptides\n",
    "from collections import Counter\n",
    "\n",
    "#Select one of the proteins containing many peptides\n",
    "protein_names = sequence_df.loc[Q_test.index].groupby('Protein').count().sort_values(by = 'Peptide')[::-1][:40]\n",
    "protein_names = protein_names.index\n",
    "protein_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Plots for top 10 test proteins\n",
    "for i in range(10):\n",
    "    createJointPlots(protein_names[i])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
